### ISSUE TITLE
False-positive for implementing Callable members with static/class methods

### ISSUE THREAD

**Issue by ivan-klass** [microsoft/pyright#7512]
```python3
from typing import Callable


class Base:

    callback: Callable[[str], int]

    def internal(self, a: str, b: str) -> int:
        """An example of internal self.callback access"""
        return max(self.callback(a), self.callback(b))


def external(b: Base, arg: str):
    b.callback(arg)


class ValidOverrideStatic(Base):
    @staticmethod  # it's Callable[[str], int] for both external and internal access
    def callback(arg: str, /) -> int:
        return len(arg)


class ValidOverrideClassLevel(Base):
    @classmethod  # it's Callable[[str], int] for both external and internal access
    def callback(cls, arg: str, /) -> int:
        return len(arg)


# this works
for instance in [ValidOverrideStatic(), ValidOverrideClassLevel()]:
    external(instance, "arg")
    instance.internal("arg1", "arg2")
```

Actual behaviour: 
```
Method "callback" overrides class "Base" in an incompatible manner
  Base method is declared as an instance method but override is not  (reportIncompatibleMethodOverride)

Method "callback" overrides class "Base" in an incompatible manner
  Base method is declared as an instance method but override is not
  Positional parameter count mismatch; base method has 1, but override has 2  (reportIncompatibleMethodOverride)
```

**Comment by erictraut** [comment]
Pyright's behavior here is correct. This isn't a bug.

You've defined an instance variable named `callback` in the base class, but you're attempting to override it in subclasses with a class variable. Callables that are instance variables act differently from callables that are class variables, so the distinction is important. The distinction between `@staticmethod` and `@classmethod` is also important, so the two cannot be used interchangeably as overrides.

If your intent is for all subclasses to implement a static method, I recommend using an ABC like this:

```python
class Base(ABC):
    @staticmethod
    @abstractmethod
    def callback(arg: str, /) -> int:
        raise NotImplementedError
```

If you don't care so much about override type safety, you can disable pyright's `reportIncompatibleMethodOverride` check or suppress it on a case-by-case basis.

**Comment by ivan-klass** [comment]
@erictraut thank you.

>If your intent is for all subclasses to implement a static method,

The intent was not to restrict subclasses in the way they define that `callback`.

Field or bound/static/class method - from the first look all behave the same externally and internally - i will appreciate if there  are any simple proofs and examples they behave differently (i.e. break Liskov principle) - except the tricky inspections/reverse-engineering, like `dict(instance)`

Thanks!

**Comment by erictraut** [comment]
> from the first look all behave the same externally and internally

It depends on whether you access them from a class object or an instance of the class.

```python
from typing import Callable

class A:
    def cb(self, a: str) -> int:
        return int(a)

class B:
    cb: Callable[[str], int]

    def __init__(self):
        self.cb = lambda a: int(a)

class C:
    cb: Callable[[str], int] = lambda a: int(a)

class D:
    @staticmethod
    def cb(a: str) -> int:
        return int(a)

class E:
    @classmethod
    def cb(cls, a: str) -> int:
        return int(a)

A.cb("1")  # Runtime exception
B.cb("1")  # Runtime exception
C.cb("1")  # OK
D.cb("1")  # OK
E.cb("1")  # OK

A().cb("1")  # OK
B().cb("1")  # OK
C().cb("1")  # Runtime exception
D().cb("1")  # OK
E().cb("1")  # OK
```

**Comment by ivan-klass** [comment]
@erictraut in your example there's no inheritance and access is on class-level, not instance level. 
For class member type definitions in potentially could be `ClassVar[Callable[[str], int]` then

**Comment by ivan-klass** [comment]
I just want to understand whether it's my misunderstanding or it's a tool limitation and we can't annotate like that just because tool doesn't support that

**Comment by erictraut** [comment]
My sample is meant to show that these different ways of declaring `cb` result in different behaviors. If pyright were to allow you to use one of these methods in a base class but a different method in a subclass, it would violate the LSP. This is why pyright forces you to be consistent across a class hierarchy.

**Comment by ivan-klass** [comment]
> these different ways of declaring cb result in different behaviors

yes, but that's a general difference, not a violation of this exact type annotation which override examples seems legit.

Difference is only when accessing on class level, while member annotation is only for instance level definitions, unless it's an explicit `ClassVar`

```python3

class MyType:
    my_member: int
    c_member: ClassVar[str]

def access_class(t: type[MyType]) -> int:  
    print(t.c_member) # this is fine
    return t.my_member # no one expects this to type check
```

But it's interesting that the following broken code is green

```python3
from typing import Callable

class Example:
    cb: Callable[[str], int] = lambda a: int(a)
```

> it would violate the LSP

Sounds like a tooling limitation then. So going on with `  # pyright: ignore[reportIncompatibleMethodOverride]` would be the option.

That's sad but fine - it's technically impossible to **statically** type-check a language where types and their shapes are created **dynamically** - custom metaclasses, decorators, descriptors...

**Comment by ivan-klass** [comment]
Workaround:
```python3

class Base:
    callback: Callable[[str], int]


class Impl(Base):
    
    @staticmethod
    def _callback(arg: str) -> int:
        return len(arg)
    
    callback: Callable[[str], int] = _callback
    
```
